File: fin/php/models/Order.php
<?php
/**
 * Order Model - نموذج الطلب
 * 
 * تم التصحيح:
 * - تصحيح أسماء الأعمدة لتتوافق مع Schema
 * - إصلاح مشكلة LIMIT/OFFSET مع PDO
 * - تحسين معالجة الأخطاء
 */

require_once __DIR__ . '/../config/database.php';
require_once __DIR__ . '/Cart.php';
require_once __DIR__ . '/Product.php';

class Order {
    private string $table = 'orders';
    private PDO $db;

    public function __construct() {
        $this->db = db();
    }

    /**
     * البحث عن طلب بالـ ID
     */
    public function findById($id): ?array {
        try {
            $stmt = $this->db->prepare("SELECT * FROM {$this->table} WHERE id = ?");
            $stmt->execute([$id]);
            $order = $stmt->fetch();

            if ($order) {
                $order['items'] = $this->getItems($id);
                // تحويل عنوان الشحن من JSON
                if (!empty($order['shipping_address'])) {
                    $order['shipping_address'] = json_decode($order['shipping_address'], true);
                }
            }

            return $order ?: null;
        } catch (PDOException $e) {
            return null;
        }
    }

    /**
     * البحث عن طلب برقم الطلب
     */
    public function findByOrderNumber(string $orderNumber): ?array {
        try {
            $stmt = $this->db->prepare("SELECT * FROM {$this->table} WHERE order_number = ?");
            $stmt->execute([$orderNumber]);
            $order = $stmt->fetch();

            if ($order) {
                $order['items'] = $this->getItems($order['id']);
                if (!empty($order['shipping_address'])) {
                    $order['shipping_address'] = json_decode($order['shipping_address'], true);
                }
            }

            return $order ?: null;
        } catch (PDOException $e) {
            return null;
        }
    }

    /**
     * الحصول على عناصر الطلب
     */
    public function getItems(int $orderId): array {
        try {
            // تصحيح: استخدام أسماء الأعمدة الصحيحة من Schema
            $sql = "SELECT oi.*, p.image, p.slug
                    FROM order_items oi
                    LEFT JOIN products p ON oi.product_id = p.id
                    WHERE oi.order_id = ?";

            $stmt = $this->db->prepare($sql);
            $stmt->execute([$orderId]);
            return $stmt->fetchAll();
        } catch (PDOException $e) {
            return [];
        }
    }

    /**
     * الحصول على طلبات المستخدم
     * تم التصحيح: استخدام bindParam بشكل صحيح
     */
    public function getByUser(int $userId, int $page = 1, int $perPage = 10): array {
        try {
            $page = max(1, (int)$page);
            $perPage = max(1, min(100, (int)$perPage));
            $offset = ($page - 1) * $perPage;

            // عد الإجمالي
            $countStmt = $this->db->prepare("SELECT COUNT(*) FROM {$this->table} WHERE user_id = ?");
            $countStmt->execute([$userId]);
            $total = (int)$countStmt->fetchColumn();

            // الاستعلام الرئيسي
            $sql = "SELECT * FROM {$this->table}
                    WHERE user_id = :userid
                    ORDER BY created_at DESC
                    LIMIT :limit OFFSET :offset";

            $stmt = $this->db->prepare($sql);
            $stmt->bindValue(':userid', $userId, PDO::PARAM_INT);
            $stmt->bindValue(':limit', $perPage, PDO::PARAM_INT);
            $stmt->bindValue(':offset', $offset, PDO::PARAM_INT);
            $stmt->execute();
            $orders = $stmt->fetchAll();

            // جلب العناصر لكل طلب
            foreach ($orders as &$order) {
                $order['items'] = $this->getItems($order['id']);
                if (!empty($order['shipping_address'])) {
                    $order['shipping_address'] = json_decode($order['shipping_address'], true);
                }
            }

            return [
                'orders' => $orders,
                'total' => $total
            ];

        } catch (PDOException $e) {
            return ['orders' => [], 'total' => 0];
        }
    }

    /**
     * إنشاء طلب جديد
     * تم التصحيح: استخدام أسماء الأعمدة الصحيحة من Schema
     */
    public function create(int $userId, array $data): array {
        $cartModel = new Cart();
        $productModel = new Product();

        try {
            $cart = $cartModel->getByUser($userId);
            
            if (empty($cart['items'])) {
                return ['success' => false, 'message' => 'السلة فارغة'];
            }

            // التحقق من الحد الأدنى للطلب
            $minOrderValue = defined('MIN_ORDER_VALUE') ? MIN_ORDER_VALUE : 0;
            if ($cart['subtotal'] < $minOrderValue) {
                $currencySymbol = defined('CURRENCY_SYMBOL') ? CURRENCY_SYMBOL : 'SAR';
                return ['success' => false, 'message' => "الحد الأدنى للطلب {$minOrderValue} {$currencySymbol}"];
            }

            // التحقق من المخزون
            foreach ($cart['items'] as $item) {
                if (!$productModel->checkStock($item['product_id'], $item['quantity'], $item['variant_id'])) {
                    return ['success' => false, 'message' => 'المنتج "' . $item['name'] . '" غير متوفر بالكمية المطلوبة'];
                }
            }

            $totals = $cartModel->calculateTotals($userId, $data['coupon_code'] ?? null);
            $orderNumber = 'ORD-' . date('Ymd') . '-' . strtoupper(substr(uniqid(), -6));

            $this->db->beginTransaction();

            // تصحيح: استخدام أسماء الأعمدة الصحيحة من Schema
            // shipping_address هو JSON في Schema وليس أعمدة منفصلة
            $shippingAddress = json_encode([
                'name' => $data['shipping_name'] ?? '',
                'phone' => $data['shipping_phone'] ?? '',
                'city' => $data['shipping_city'] ?? '',
                'street' => $data['shipping_address'] ?? '',
                'notes' => $data['notes'] ?? ''
            ], JSON_UNESCAPED_UNICODE);

            $sql = "INSERT INTO {$this->table}
                    (order_number, user_id, status, payment_status, payment_method, 
                     subtotal, discount, tax, shipping_cost, total, coupon_code, shipping_address, notes)
                    VALUES (?, ?, 'pending', 'pending', ?, ?, ?, ?, ?, ?, ?, ?, ?)";
            
            $stmt = $this->db->prepare($sql);
            $stmt->execute([
                $orderNumber,
                $userId,
                $data['payment_method'] ?? 'cash',
                $totals['subtotal'],
                $totals['discount'],
                $totals['tax'],
                $totals['shipping'],
                $totals['total'],
                $data['coupon_code'] ?? null,
                $shippingAddress,
                $data['notes'] ?? null
            ]);
            
            $orderId = (int)$this->db->lastInsertId();

            // إضافة عناصر الطلب
            // تصحيح: استخدام أسماء الأعمدة الصحيحة (name بدلاً من product_name)
            $itemSql = "INSERT INTO order_items (order_id, product_id, variant_id, name, price, quantity, total, options)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?)";
            $itemStmt = $this->db->prepare($itemSql);
            
            foreach ($cart['items'] as $item) {
                $options = $item['variant_name'] ? json_encode(['variant' => $item['variant_name']], JSON_UNESCAPED_UNICODE) : null;
                
                $itemStmt->execute([
                    $orderId,
                    $item['product_id'],
                    $item['variant_id'],
                    $item['name'],
                    $item['unit_price'],
                    $item['quantity'],
                    $item['total'],
                    $options
                ]);
                
                // تقليل المخزون
                $productModel->reduceStock($item['product_id'], $item['quantity'], $item['variant_id']);
            }

            // تحديث استخدام الكوبون
            if (!empty($data['coupon_code'])) {
                $this->db->prepare("UPDATE coupons SET used_count = used_count + 1 WHERE code = ?")->execute([$data['coupon_code']]);
            }

            // تفريغ السلة
            $cartModel->clear($userId);
            
            $this->db->commit();

            return ['success' => true, 'order' => $this->findById($orderId)];

        } catch (Exception $e) {
            if ($this->db->inTransaction()) {
                $this->db->rollBack();
            }
            return ['success' => false, 'message' => 'حدث خطأ أثناء إنشاء الطلب'];
        }
    }

    /**
     * إلغاء الطلب
     */
    public function cancel(int $orderId, int $userId): array {
        try {
            $order = $this->findById($orderId);

            if (!$order || $order['user_id'] != $userId) {
                return ['success' => false, 'message' => 'الطلب غير موجود'];
            }

            // تصحيح: التحقق من حالات يمكن إلغاؤها
            $cancellableStatuses = ['pending', 'processing'];
            if (!in_array($order['status'], $cancellableStatuses)) {
                return ['success' => false, 'message' => 'لا يمكن إلغاء هذا الطلب'];
            }

            $productModel = new Product();
            $this->db->beginTransaction();

            // تحديث حالة الطلب
            $this->db->prepare("UPDATE {$this->table} SET status = 'cancelled', cancelled_at = NOW() WHERE id = ?")->execute([$orderId]);
            
            // استعادة المخزون
            foreach ($order['items'] as $item) {
                $productModel->restoreStock($item['product_id'], $item['quantity'], $item['variant_id']);
            }

            $this->db->commit();

            return ['success' => true, 'order' => $this->findById($orderId)];

        } catch (Exception $e) {
            if ($this->db->inTransaction()) {
                $this->db->rollBack();
            }
            return ['success' => false, 'message' => 'حدث خطأ أثناء إلغاء الطلب'];
        }
    }

    /**
     * تحديث حالة الطلب (للمدير)
     */
    public function updateStatus(int $orderId, string $status): bool {
        try {
            $validStatuses = ['pending', 'processing', 'shipped', 'delivered', 'cancelled', 'refunded'];
            if (!in_array($status, $validStatuses)) {
                return false;
            }

            $sql = "UPDATE {$this->table} SET status = ?";
            $params = [$status];

            // إضافة التوقيت المناسب
            if ($status === 'shipped') {
                $sql .= ", shipped_at = NOW()";
            } elseif ($status === 'delivered') {
                $sql .= ", delivered_at = NOW()";
            } elseif ($status === 'cancelled') {
                $sql .= ", cancelled_at = NOW()";
            }

            $sql .= ", updated_at = NOW() WHERE id = ?";
            $params[] = $orderId;

            $stmt = $this->db->prepare($sql);
            return $stmt->execute($params);

        } catch (PDOException $e) {
            return false;
        }
    }
}

File: lib/models/order.dart
import 'cart_item.dart';

/// نموذج الطلب
/// تم التصحيح: مطابقة حقل shipping_address مع PHP (JSON)
class Order {
  final int id;
  final String? orderNumber;
  final int userId;
  final List<CartItem> items;
  final double subtotal;
  final double shipping;
  final double tax;
  final double discount;
  final double total;
  final String status;
  final String paymentStatus;
  final ShippingAddress? shippingAddress; // تم التصحيح: كائن بدلاً من حقول منفصلة
  final String? paymentMethod;
  final String? notes;
  final String? couponCode;
  final DateTime createdAt;
  final DateTime? shippedAt;
  final DateTime? deliveredAt;
  final DateTime? updatedAt;

  Order({
    required this.id,
    this.orderNumber,
    required this.userId,
    this.items = const [],
    required this.subtotal,
    this.shipping = 0,
    this.tax = 0,
    this.discount = 0,
    required this.total,
    required this.status,
    this.paymentStatus = 'pending',
    this.shippingAddress,
    this.paymentMethod,
    this.notes,
    this.couponCode,
    required this.createdAt,
    this.shippedAt,
    this.deliveredAt,
    this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    List<CartItem> orderItems = [];
    if (json['items'] != null && json['items'] is List) {
      orderItems = (json['items'] as List)
          .map((item) => CartItem.fromJson(item))
          .toList();
    }

    // تم التصحيح: قراءة shipping_address ككائن JSON
    ShippingAddress? address;
    if (json['shipping_address'] != null) {
      if (json['shipping_address'] is Map) {
        address = ShippingAddress.fromJson(json['shipping_address']);
      } else if (json['shipping_address'] is String) {
        // محاولة تحويل النص إلى JSON
        try {
          final decoded = json['shipping_address'];
          if (decoded is Map) {
            address = ShippingAddress.fromJson(decoded as Map<String, dynamic>);
          }
        } catch (e) {
          // تجاهل الخطأ
        }
      }
    }

    return Order(
      id: _parseInt(json['id']),
      orderNumber: json['order_number']?.toString(),
      userId: _parseInt(json['user_id']),
      items: orderItems,
      subtotal: _parseDouble(json['subtotal']),
      shipping: _parseDouble(json['shipping']),
      tax: _parseDouble(json['tax']),
      discount: _parseDouble(json['discount']),
      total: _parseDouble(json['total']),
      status: json['status']?.toString() ?? 'pending',
      paymentStatus: json['payment_status']?.toString() ?? 'pending',
      shippingAddress: address,
      paymentMethod: json['payment_method']?.toString(),
      notes: json['notes']?.toString(),
      couponCode: json['coupon_code']?.toString(),
      createdAt: _parseDateTime(json['created_at']),
      shippedAt: json['shipped_at'] != null ? _parseDateTime(json['shipped_at']) : null,
      deliveredAt: json['delivered_at'] != null ? _parseDateTime(json['delivered_at']) : null,
      updatedAt: json['updated_at'] != null ? _parseDateTime(json['updated_at']) : null,
    );
  }

  static int _parseInt(dynamic value, [int defaultValue = 0]) {
    if (value == null) return defaultValue;
    if (value is int) return value;
    if (value is String) return int.tryParse(value) ?? defaultValue;
    return defaultValue;
  }

  static double _parseDouble(dynamic value, [double defaultValue = 0.0]) {
    if (value == null) return defaultValue;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? defaultValue;
    return defaultValue;
  }

  static DateTime _parseDateTime(dynamic value) {
    if (value == null) return DateTime.now();
    if (value is DateTime) return value;
    if (value is String) return DateTime.tryParse(value) ?? DateTime.now();
    return DateTime.now();
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'order_number': orderNumber,
      'user_id': userId,
      'items': items.map((item) => item.toJson()).toList(),
      'subtotal': subtotal,
      'shipping': shipping,
      'tax': tax,
      'discount': discount,
      'total': total,
      'status': status,
      'payment_status': paymentStatus,
      'shipping_address': shippingAddress?.toJson(),
      'payment_method': paymentMethod,
      'notes': notes,
      'coupon_code': couponCode,
      'created_at': createdAt.toIso8601String(),
      'shipped_at': shippedAt?.toIso8601String(),
      'delivered_at': deliveredAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  Order copyWith({
    int? id,
    String? orderNumber,
    int? userId,
    List<CartItem>? items,
    double? subtotal,
    double? shipping,
    double? tax,
    double? discount,
    double? total,
    String? status,
    String? paymentStatus,
    ShippingAddress? shippingAddress,
    String? paymentMethod,
    String? notes,
    String? couponCode,
    DateTime? createdAt,
    DateTime? shippedAt,
    DateTime? deliveredAt,
    DateTime? updatedAt,
  }) {
    return Order(
      id: id ?? this.id,
      orderNumber: orderNumber ?? this.orderNumber,
      userId: userId ?? this.userId,
      items: items ?? this.items,
      subtotal: subtotal ?? this.subtotal,
      shipping: shipping ?? this.shipping,
      tax: tax ?? this.tax,
      discount: discount ?? this.discount,
      total: total ?? this.total,
      status: status ?? this.status,
      paymentStatus: paymentStatus ?? this.paymentStatus,
      shippingAddress: shippingAddress ?? this.shippingAddress,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      notes: notes ?? this.notes,
      couponCode: couponCode ?? this.couponCode,
      createdAt: createdAt ?? this.createdAt,
      shippedAt: shippedAt ?? this.shippedAt,
      deliveredAt: deliveredAt ?? this.deliveredAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  String get statusText {
    switch (status) {
      case 'pending':
        return 'قيد الانتظار';
      case 'processing':
        return 'قيد المعالجة';
      case 'shipped':
        return 'تم الشحن';
      case 'delivered':
        return 'تم التوصيل';
      case 'cancelled':
        return 'ملغي';
      default:
        return status;
    }
  }

  String get paymentStatusText {
    switch (paymentStatus) {
      case 'pending':
        return 'في انتظار الدفع';
      case 'paid':
        return 'تم الدفع';
      case 'failed':
        return 'فشل الدفع';
      case 'refunded':
        return 'تم الاسترداد';
      default:
        return paymentStatus;
    }
  }
}

/// نموذج عنوان الشحن
class ShippingAddress {
  final String name;
  final String phone;
  final String address;
  final String? city;
  final String? state;
  final String? postalCode;
  final String? country;

  ShippingAddress({
    required this.name,
    required this.phone,
    required this.address,
    this.city,
    this.state,
    this.postalCode,
    this.country,
  });

  factory ShippingAddress.fromJson(Map<String, dynamic> json) {
    return ShippingAddress(
      name: json['name']?.toString() ?? '',
      phone: json['phone']?.toString() ?? '',
      address: json['address']?.toString() ?? '',
      city: json['city']?.toString(),
      state: json['state']?.toString(),
      postalCode: json['postal_code']?.toString(),
      country: json['country']?.toString(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'phone': phone,
      'address': address,
      'city': city,
      'state': state,
      'postal_code': postalCode,
      'country': country,
    };
  }
}
